File: 061-RotateList.py
# Definition for singly-linked list.
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head == None:
            return None

        start = head
        stack_t = []
        while start.next:
            stack_t.append(start)
            start= start.next
        stack_t.append(start)
        k = k % len(stack_t)
        if k == 0:
            return head
        stack_t[-1].next = head
        head = stack_t[-k]
        stack_t[-(k+1)].next = None

        return head


if __name__ == '__main__':

    #inputs = raw_input('input the N :  ')

    head = None
    for i in reversed(range(int(1))):
        t = ListNode(i+1)
        t.next = head
        head = t

    A = Solution()
    k = 2
    r = A.rotateRight(head,k)

    print 'Result:',
    while r:
        print r.val,'->',
        r = r.next
    print 'NULL'

File: 061_RotateList.js
//Definition for singly-linked list.
function ListNode(val){
     this.val = val;
     this.next = null;
};

/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    //Runtime: 72 ms, faster than 65.23% of JavaScript online submissions for Rotate List.
    //Memory Usage: 35.7 MB, less than 17.39% of JavaScript online submissions for Rotate List.
    if (head===null || head.next===null||k<=0){
        return(head);
    };
    var pre_head= head;
    var node_stack =[];
    while (head.next!=null){
        node_stack.push(head);
        head=head.next;
    };
    node_stack.push(head);
    var l= node_stack.length;
    k = k%l;
    if (k===0){
        return(pre_head);
    };
    node_stack[l-1].next= pre_head;
    node_stack[l-k-1].next = null;
    return(node_stack[l-k]);
};

var n1= new ListNode(1);
var n2= new ListNode(2);
var n3= new ListNode(3);
var n4= new ListNode(4);
var n5= new ListNode(5);
n1.next=n2;
//n2.next=n3;
//n3.next=n4;
//n4.next=n5;
var k=2;

console.log(rotateRight(n1,k).val);
console.log('-----------the end!----------------');

